### PR TITLE
Tab Component

### PR DESCRIPTION
**Pull request by yomed**
## Description

- adds initial tab component
- has minor changes to common utils and test utils

## Context

This component follows a similar pattern to the carousel in terms of selection. A heading (and a panel, in case of non-fake use case) is selected via the `index` property on the root element. I started out using a boolean attribute on the heading itself, like `<ebay-tab-heading selected/>`, but ran into lots of issues. A single heading must be selected at all times, so the normalization gets very messy in the boolean case.

## References

#258 

## Screenshots

screen shot 2018-08-02 at 4 48 49 pm
screen shot 2018-08-02 at 4 48 55 pm
screen shot 2018-08-02 at 4 48 59 pm
screen shot 2018-08-02 at 4 49 03 pm

### CODE REVIEW

File: marko.json [hunk 1]
```diff
@@ -192,5 +192,27 @@
             "@selected": "boolean"
         }
     },
-    "<ebay-select-option>": {}
+    "<ebay-select-option>": {},
+    "<ebay-tab>": {
+        "renderer": "./src/components/ebay-tab/index.js",
+        "transformer": "./src/common/transformer/index.js",
+        "@*": "expression",
+        "@html-attributes": "expression",
```
Comment (senthilp) [inline marko.json]: Do we really need `@html-attributes`? `@*` should be sufficient right? This is a general comment, not specific to this component.

Comment (yomed) [reply marko.json]: This was introduced to support dynamic pass-through attributes. More details here: https://github.com/eBay/ebayui-core/pull/154

File: src/components/ebay-tab/template.marko [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                </div>
+            </div>
+        </div>
+    </if>
+    <else>
+        <div w-id="items" class="tabs__items" role="tablist">
+            <for(i, item in data.items)>
+                <div
+                    aria-controls="${widget.id}-tabpanel-${i}"
+                    id="${widget.id}-default-tab-${i}"
```
Comment (sendlo) [inline src/components/ebay-tab/template.marko]: Minor - why are they all called "default-tab" (plus id and item number)?  I would see this and think whatever tab I'm looking at is the default.

File: src/components/ebay-tab/index.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+/**
+ * Common processing of index change via both UI and API
+ * @param {Number} index
+ */
+function processStateChange(index) {
+    if (index >= 0 && index < this.state.items.length && index !== this.state.index) {
+        this.setState('index', index);
+        emitAndFire(this, 'tab-select', { index });
```
Comment (sendlo) [inline src/components/ebay-tab/index.js]: documentation says it emits the original event too

File: src/components/ebay-tab/template.marko (hunk 2 above)
Comment (yomed) [reply src/components/ebay-tab/template.marko]: Thanks, I'll update. This was from the Skin example.

File: src/components/ebay-tab/index.js (hunk 3 above)
Comment (sendlo) [reply src/components/ebay-tab/index.js]: To clarify:
https://github.com/eBay/ebayui-core/tree/258-tab-component/src/components/ebay-tab#ebay-tab-events

It looks like the documentation says the original event is passed as the data.  I think it makes sense to pass the original event and the element, similar to what we do in pagination.

Comment (yomed) [reply src/components/ebay-tab/index.js]: We don't currently send the `originalEvent` for cases that can be triggered both programmatically and by UI events. So I'm going to leave it out of this component until we determine the general pattern.

Comment (senthilp) [reply src/components/ebay-tab/index.js]: Agree. For now, we will skip sending `originalEvent ` for programmatically triggered events.

**Comment by DylanPiercey** [comment]
I think everything looks good here. My only reservation is the name `ebay-tab-item` is a little generic. I'd personally prefer something like `ebay-tab-heading`. If every one is happy with the current naming though then that's fine with me 😄.

**Comment by sendlo** [comment]
I like Dylan's recommendation of changing it to ebay-tab-heading too.  (I actually like ebay-tab-title.) Tab item could be confused with the content.

**Pull request merged**